File: src/server/pipelines/consumer.go
package main

import (
    "context"
    "fmt"
    "log"
    "os"
    "github.com/segmentio/kafka-go"
)

const (
    kafkaBroker = "127.0.0.1:9092"
    topic       = "data_pipeline"
    groupID     = "data_pipeline_group"
)

func main() {
    r := kafka.NewReader(kafka.ReaderConfig{
        Brokers:  []string{kafkaBroker},
        Topic:    topic,
        GroupID:  groupID,
        MinBytes: 10e3, // 10KB
        MaxBytes: 10e6, // 10MB
    })

    file, err := os.OpenFile("data/output.txt", os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
    if err != nil {
        log.Fatal("could not open output file " + err.Error())
    }
    defer file.Close()

    for {
        m, err := r.FetchMessage(context.Background())
        if err != nil {
            log.Fatal("could not read message " + err.Error())
        }
        fmt.Printf("Received: %s\n", string(m.Value))
        if _, err := file.WriteString(fmt.Sprintf("%s\n", m.Value)); err != nil {
            log.Fatal("could not write to output file " + err.Error())
        }

        if err := r.CommitMessages(context.Background(), m); err != nil {
            log.Fatal("could not commit message " + err.Error()))
        }
    }
}

File: src/server/pipelines/consumer.rs
use kafka::consumer::{Consumer, FetchOffset, GroupOffsetStorage};
use std::time::Duration;
use std::fs::OpenOptions;
use std::io::Write;

const KAFKA_BROKER: &str = "127.0.0.1:9092";
const TOPIC: &str = "data_pipeline";
const GROUP_ID: &str = "data_pipeline_group";

fn main() {
    let mut consumer = Consumer::from_hosts(vec![KAFKA_BROKER.to_owned()])
        .with_topic(TOPIC.to_owned())
        .with_group(GROUP_ID.to_owned())
        .with_fallback_offset(FetchOffset::Earliest)
        .with_offset_storage(GroupOffsetStorage::Kafka)
        .create()
        .unwrap();

    let mut file = OpenOptions::new().append(true).open("data/output.txt").unwrap();

    loop {
        for ms in consumer.poll().unwrap().iter() {
            for m in ms.messages() {
                let chunk = String::from_utf8(m.value.to_vec()).unwrap();
                println!("Received: {}", chunk);
                writeln!(file, "{}", chunk).unwrap();
            }
            consumer.consume_messageset(ms).unwrap();
        }
        consumer.commit_consumed().unwrap();
        std::thread::sleep(Duration::from_secs(1)); // Polling interval
    }
}
